pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,552 @@
+		return 0, errs.ErrGCOnInvalidKeyspace.GenWithStackByArgs(keyspaceID)
+	}
+
+	return keyspaceID, nil
+}
+
+// AdvanceGCSafePoint tries to advance the GC safe point to the given target. If the target is less than the current
+// value or greater than the txn safe point, it returns an error.
+//
+// WARNING: This method is only used to manage the GC procedure, and should never be called by code that doesn't
+// have the responsibility to manage GC. It can only be called on NullKeyspace or keyspaces with keyspace level GC
+// enabled.
+func (m *GCStateManager) AdvanceGCSafePoint(keyspaceID uint32, target uint64) (oldGCSafePoint uint64, newGCSafePoint uint64, err error) {
+	keyspaceID, err = m.redirectKeyspace(keyspaceID, false)
+	if err != nil {
+		return
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.advanceGCSafePointImpl(keyspaceID, target, false)
+}
+
+func (m *GCStateManager) advanceGCSafePointImpl(keyspaceID uint32, target uint64, compatible bool) (oldGCSafePoint uint64, newGCSafePoint uint64, err error) {
+	newGCSafePoint = target
+
+	err = m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		var err1 error
+		oldGCSafePoint, err1 = m.gcMetaStorage.LoadGCSafePoint(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+		if target < oldGCSafePoint {
+			if compatible {
+				// When in compatible mode, trying to update the safe point to a smaller value fails silently, returning
+				// the actual value. There exist some use cases that fetches the current value by passing zero.
+				log.Warn("deprecated API `UpdateGCSafePoint` is called with invalid argument",
+					zap.Uint64("currentGCSafePoint", oldGCSafePoint), zap.Uint64("attemptedGCSafePoint", target))
+				newGCSafePoint = oldGCSafePoint
+				return nil
+			}
+			// Otherwise, return error to reject the operation explicitly.
+			return errs.ErrDecreasingGCSafePoint.GenWithStackByArgs(oldGCSafePoint, target)
+		}
+		txnSafePoint, err1 := m.gcMetaStorage.LoadTxnSafePoint(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+		if target > txnSafePoint {
+			return errs.ErrGCSafePointExceedsTxnSafePoint.GenWithStackByArgs(oldGCSafePoint, target, txnSafePoint)
+		}
+
+		return wb.SetGCSafePoint(keyspaceID, target)
+	})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if keyspaceID == constant.NullKeyspaceID {
+		gcSafePointGauge.WithLabelValues("gc_safepoint").Set(float64(target))
+	}
+
+	return
+}
+
+// AdvanceTxnSafePoint tries to advance the txn safe point to the given target.
+//
+// Returns a struct AdvanceTxnSafePointResult, which contains the old txn safe point, the target, and the new
+// txn safe point it finally made it to advance to. If there's something blocking the txn safe point from being
+// advanced to the given target, it may finally be advanced to a smaller value or remains the previous value, in which
+// case the BlockerDescription field of the AdvanceTxnSafePointResult will be set to a non-empty string describing
```

Comment:
Do blocker and barrier have the same meaning? If so, how about unifying them?

Reply:
Apart from the GC barriers, The *blocker* can also be a *TiDB min start TS* from an earlier version of TiDB that didn't follow up the refactoring. It need to be considered for compatibility concerns.